Add exclude_patterns feature based on sphinx Matcher

n-peugnet opened the pull request:
Another option similar to #41 to ignore some changes in dates. This is especially useful for ignoring files in dependencies. This way the date does not change if an ignored dependent file is updated.

My current use case would be to ignore the `docutils.conf` file, as I it is added as a dependency by Sphinx to all documents, but I only want to reflect content changes in the "last updated date".
The current implementation uses [`fnmatch`](https://docs.python.org/3/library/fnmatch.html#fnmatch.fnmatch) so some kinds of wildcard patterns are accepted. In my case I would use the following in `conf.py`:

```py
git_ignored_files = [
    '*/docutils.conf',
]
```

As for #41, the name of the config key can be changed, and docs will follow after a first review of the patch, just to see if you are willing to accept these changes.

mgeier commented on src/sphinx_last_updated_by_git.py:
```diff
@@ -114,6 +116,14 @@ def parse_log(stream, requested_files, git_dir, file_dates):
+def is_file_ignored(file: Path, ignored_files: List[str]) -> bool:
+    """Check if a file matches a list of ignored files using fnmatch"""
+    for ignored_file in ignored_files:
+        if fnmatch(str(file), ignored_file):
```
I'm not really familiar with `fnmatch()`, but couldn't you use `pathlib` here as well?

https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.match

mgeier replied:
Now that I'm thinking about `exclude_patterns`, maybe we could use `sphinx.util.matching.Matcher` to get the exact same behavior?

mgeier commented on src/sphinx_last_updated_by_git.py:
```diff
@@ -114,6 +116,14 @@ def parse_log(stream, requested_files, git_dir, file_dates):
             break
 
 
+def is_file_ignored(file: Path, ignored_files: List[str]) -> bool:
```
I'm not against using type annotations, but wouldn't it be better to check them in CI?

I'm not really familiar with type annotations, but according to https://docs.python.org/3/library/typing.html#typing.List, `Iterable[str]` is preferred. Or is that about not allowing `str` (https://github.com/python/typing/issues/256)?

n-peugnet replied:
> I'm not against using type annotations, but wouldn't it be better to check them in CI?

I agree that it would be better, but I don't think that it is a problem if it is not yet the case.
[The runtime does not check them at all](https://docs.python.org/3/library/typing.html) so it is not that much of a problem if they are not perfect:

> The Python runtime does not enforce function and variable type annotations. They can be used by third party tools such as type checkers, IDEs, linters, etc.

Concerning the following:

> I'm not really familiar with type annotations, but according to https://docs.python.org/3/library/typing.html#typing.List, `Iterable[str]` is preferred. Or is that about not allowing `str` ([python/typing#256](https://github.com/python/typing/issues/256))?

You are entirely right, `Iterable` makes a lot more sense. I am not extremely familiar with Python type annotations either :sweat_smile:

mgeier commented on src/sphinx_last_updated_by_git.py:
```diff
@@ -114,6 +116,14 @@ def parse_log(stream, requested_files, git_dir, file_dates):
             break
 
 
+def is_file_ignored(file: Path, ignored_files: List[str]) -> bool:
```
Since both uses of this function are within the same function, you could move it into that function as a closure over `app.config.git_ignored_files`, which would avoid the second function argument.

mgeier commented on src/sphinx_last_updated_by_git.py:
```diff
@@ -114,6 +116,14 @@ def parse_log(stream, requested_files, git_dir, file_dates):
             break
 
 
+def is_file_ignored(file: Path, ignored_files: List[str]) -> bool:
```
Wouldn't the name `file_is_ignored()` be better to read?

mgeier replied:
In the Sphinx source code, the `Matcher` is simply called `excluded()`.

n-peugnet commented:
@mgeier I updated this PR based on your comments (thank you for your great reviews by the way) and added tests and docs. I can squash the commits and fix the anticipated conflicts if needed.

mgeier commented on README.rst:
```diff
@@ -39,6 +39,13 @@ Options
+      a list of exclusion patterns to the configuration option
+      ``git_exclude_patterns``.
+      This can also be used to ignore the changes of dependencies in the last
+      updated date.
```
This sentence is not quite clear to me. There are two ways I might understand that:

* if a source file is ignored, its dependencies are ignored as well
* specific dependencies can be ignored just as well as source files

I think the first point makes sense, I'm not sure if we should explicitly mention this?
I guess you meant the second point?
Can you please re-phrase the sentence to make it clearer?

n-peugnet replied:
Yes I meant the second point, I also realized later that this sentence was not very clear.

mgeier commented on tests/test_example_repo.py:
```diff
@@ -123,3 +123,40 @@ def test_no_git_no_warning(capsys):
+def test_exclude_patterns_glob():
+    data = run_sphinx(
+        'repo_full',
+        git_exclude_patterns='**/*.rst',
```
This could be shortened to ...

```suggestion
        git_exclude_patterns='**.rst',
```

... but I'm not sure if that's better.
But it would behave exactly the same, right?

The test repo has no subdirectories, so even `'*.rst'` should be sufficient, but `'**.rst'` could be considered more "future-proof" if directories are added later.

It would have been good to have subdirectories in the test repo, but I don't think it's worth adding them just for that.

n-peugnet replied:
> This could be shortened to ...
>
>     git_exclude_patterns='**.rst',
>
> ... but I'm not sure if that's better. But it would behave exactly the same, right?

Based on <https://www.sphinx-doc.org/en/master/usage/configuration.html#id12>:

>  A double star ** can be used to match any sequence of characters including slashes.

And <https://github.com/sphinx-doc/sphinx/blob/5e9550c78e3421dd7dcab037021d996841178f67/sphinx/util/matching.py#L60-L65>:

```python
class Matcher:
    """A pattern matcher for Multiple shell-style glob patterns.
    Note: this modifies the patterns to work with copy_asset().
          For example, "**/index.rst" matches with "index.rst"
    """
```

Yes it would behave exactly the same.
I originally used this pattern to make sure that the `**` was working correctly, but as there is no subdirectory this is not a very effective test.

> The test repo has no subdirectories, so even `'*.rst'` should be sufficient, but `'**.rst'` could be considered more "future-proof" if directories are added later.

Maybe this is a good argument to choose `'*.rst'` instead as this way the validity of the test is less likely to change overtime.

> It would have been good to have subdirectories in the test repo, but I don't think it's worth adding them just for that.

I don't think so either. The Matcher from Sphinx is probably sufficiently tested on their side.

mgeier commented on tests/test_example_repo.py:
```diff
@@ -123,3 +123,40 @@ def test_no_git_no_warning(capsys):
+def test_exclude_patterns_deps_dates():
+    data = run_sphinx(
+        'repo_full',
+        git_exclude_patterns='*.py',
```
Since basically the same glob pattern has already been tested above, you could be a more specific here:

```suggestion
        git_exclude_patterns='example_module.py',
```

This would make it clear that `conf.py` is not involved.

mgeier reviewed (commented):
Thanks for the updates!

mgeier commented:
> I can squash the commits

No need, Github did that for me.

Thanks again for your contribution!
